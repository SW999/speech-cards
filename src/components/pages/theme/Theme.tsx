import * as React from 'react';
import { FunctionComponent, useState } from 'react';
import { ThemeItem } from '../../theme-item/ThemeItem';
import { THEMES } from '../../../constants';

import alternativeTheme from '../../../img/alternativeTheme.jpg';
import darkTheme from '../../../img/darkTheme.jpg';
import defaultTheme from '../../../img/defaultTheme.jpg';
import simpleTheme from '../../../img/simpleTheme.jpg';

const THEME_ITEMS = [
  { name: THEMES.DEFAULT, img: defaultTheme },
  { name: THEMES.DARK, img: darkTheme },
  { name: THEMES.SIMPLE, img: simpleTheme },
  { name: THEMES.ALTERNATIVE, img: alternativeTheme },
];

export const Theme: FunctionComponent = () => {
  const [active, setActive] = useState<string>(
    localStorage.getItem('speechTheme') || THEMES.DEFAULT
  );
  const onSetTheme = (name: string): void => {
    setActive(name);
    localStorage.setItem('speechTheme', name);
  };

  return (
    <>
      <h1>Select speech theme</h1>
      <p>
        Please choose a theme for displaying the speech/interview that is
        convenient for you.
      </p>
      <div className="theme-row">
        {THEME_ITEMS.map(theme => {
          return (
            <div
              key={theme.name}
              className={`theme-col${
                active === theme.name ? ' theme-col_active' : ''
              }`}
            >
              <ThemeItem
                name={theme.name}
                pic={theme.img}
                setTheme={onSetTheme}
              />
            </div>
          );
        })}
      </div>
    </>
  );
};
