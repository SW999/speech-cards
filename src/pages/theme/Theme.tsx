import React, { FunctionComponent, useState } from 'react';
import { THEMES, STORAGE_THEME_PREFIX } from '../../constants';

import darkTheme from '../../img/darkTheme.jpg';
import defaultTheme from '../../img/defaultTheme.jpg';
import simpleTheme from '../../img/simpleTheme.jpg';

const THEME_ITEMS = [
  { name: THEMES.DEFAULT, img: defaultTheme },
  { name: THEMES.DARK, img: darkTheme },
  { name: THEMES.SIMPLE, img: simpleTheme },
];

const Theme: FunctionComponent = () => {
  const [active, setActive] = useState<string>(
    localStorage.getItem(STORAGE_THEME_PREFIX) || THEMES.DEFAULT
  );
  const onSetTheme = (name: string): void => {
    if (name === active) return;

    setActive(name);
    localStorage.setItem(STORAGE_THEME_PREFIX, name);
  };

  return (
    <>
      <h1>Select speech theme</h1>
      <p>
        Please choose a theme of the&nbsp;speech/interview that is convenient
        for you.
      </p>
      <div className="theme-row">
        {THEME_ITEMS.map(theme => {
          const { name, img } = theme;
          return (
            <div
              key={name}
              className={`theme-col${
                active === name ? ' theme-col_active' : ''
              }`}
              onClick={() => onSetTheme(name)}
            >
              <img
                alt={`${name} theme`}
                className="theme"
                src={img}
                title={`${name} theme`}
              />
            </div>
          );
        })}
      </div>
    </>
  );
};

export default Theme;
