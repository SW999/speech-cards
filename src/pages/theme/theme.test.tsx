import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Theme from './Theme';
//
// const mockWindowProperty = (property, value) => {
//   const { [property]: originalProperty } = window;
//   delete window[property];
//   beforeAll(() => {
//     Object.defineProperty(window, property, {
//       value,
//       configurable: true,
//       writable: true,
//     });
//   });
//   afterAll(() => {
//     window[property] = originalProperty;
//   });
// };

describe('Theme component', () => {
  // mockWindowProperty('localStorage', {
  //   setItem: jest.fn(),
  //   getItem: jest.fn(),
  // });

  it('Theme renders correctly', () => {
    const { asFragment } = render(<Theme />);

    expect(asFragment()).toMatchSnapshot('Theme');
  });

  it('Theme not changes value on click to active', () => {
    render(<Theme />);
    fireEvent.click(screen.getByTestId('default'));

    expect(
      screen.getByTestId('default').classList.contains('theme-col_active')
    ).toBeTruthy();
  });

  it('Theme changes value on click to not active value', async () => {
    render(<Theme />);

    expect(
      screen.getByTestId('dark').classList.contains('theme-col_active')
    ).toBeFalsy();

    fireEvent.click(screen.getByTestId('dark'));

    await waitFor(() => {
      expect(
        screen.getByTestId('dark').classList.contains('theme-col_active')
      ).toBeTruthy();
    });
  });
});
