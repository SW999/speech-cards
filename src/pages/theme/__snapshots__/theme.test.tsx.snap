// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme component Theme renders correctly: Theme 1`] = `
<DocumentFragment>
  <h1>
    Select speech theme
  </h1>
  <p>
    Please choose a theme of theÂ speech/interview that is convenient for you.
  </p>
  <div
    class="theme-row"
  >
    <div
      class="theme-col theme-col_active"
      data-testid="default"
    >
      <img
        alt="default theme"
        class="theme"
        src="[object Object]"
        title="default theme"
      />
    </div>
    <div
      class="theme-col"
      data-testid="dark"
    >
      <img
        alt="dark theme"
        class="theme"
        src="[object Object]"
        title="dark theme"
      />
    </div>
    <div
      class="theme-col"
      data-testid="simple"
    >
      <img
        alt="simple theme"
        class="theme"
        src="[object Object]"
        title="simple theme"
      />
    </div>
  </div>
</DocumentFragment>
`;
